File: C#/Data-Structures/LinkedList/DoublyLinkedList.cs
using System;
using System.Collections.Generic;

/// <summary>
/// A DoublyLinkedList
/// </summary>
namespace DataStructures
{
    public class DoublyLinkedListNode<T>
    {
        // Constructs a new node with the specified value.
        public DoublyLinkedListNode(T value)
        {
            Value = value;
            Next = null;
            Prev = null;
        }
        // The node value
        public T Value { get; set; }
        // The next node in the Doubly linked list (null if last node)
        public DoublyLinkedListNode<T> Next { get; set; }
        public DoublyLinkedListNode<T> Prev { get; set; }
    }

    /// <summary>
    /// A Doubly linked list collection capable of basic operations such as 
    /// Add, Remove, Check, Search, IsEmpty, Size, Insert and Index
    /// </summary>
    public class DoublyLinkedList<T>
    {
        /// <summary>
        /// Adds the specified value to the tail of the doublylinked list
        /// </summary>
        public int Add(T item)
        {
            DoublyLinkedListNode<T> node = new DoublyLinkedListNode<T>(item);
            if (count == 0)
            {
                head = node;
                tail = head;
            }
            else
            {
                node.Prev = tail;
                tail.Next = node;
                tail = node;
            }

            return ++count;
        }

        /// <summary>
        /// Inserts the specified item at the specified index
        /// </summary>
        /// <param name="index"></param>
        /// <param name="item"></param>
        public void Insert(int index, T item)
        {
            if (index > count + 1 || (count == 0 && index > 0) || index < 0)
            {
                throw new IndexOutOfRangeException("Index out of range.");
            }
            if (count == 0 || index == count)
            {
                Add(item);
                return;
            }
            var currentIndex = 0;
            DoublyLinkedListNode<T> current = head;
            DoublyLinkedListNode<T> node = new DoublyLinkedListNode<T>(item);
            while (currentIndex <= index)
            {
                if(currentIndex == index)
                {
                    DoublyLinkedListNode<T> previous = current.Prev;
                    if (previous == null) head = node;
                    node.Prev = previous;
                    node.Next = current;
                    //previous.Next = node;
                    current.Prev = node;
                    count++;
                    return;
                }
                current = current.Next;
                currentIndex++;
            }
            return;
        }

        /// <summary>
        /// Removes a VALUE from the list if found.
        /// </summary>
        public bool Remove(T item)
        {
            var check = false;

            DoublyLinkedListNode<T> node = new DoublyLinkedListNode<T>(item);

            // Checks if only one item is in the list
            // and if the item is the same as the item to be removed
            if (count == 1 && item.Equals(head.Value))
            {
                check = true;
                head = null;
                tail = head;
            }

            // Iterates through the doublylinked list to check for the item
            while (Check(item))
            {
                check = true;

                DoublyLinkedListNode<T> current = head;
                while (current != null)
                {
                    if (current.Value.Equals(item))
                    {
                        DoublyLinkedListNode<T> previous = current.Prev;
                        DoublyLinkedListNode<T> next = current.Next;

                        if (previous == null)
                        {
                            head = next;
                            next.Prev = null;
                        }
                        else if (next == null)
                        {
                            tail = previous;
                            previous.Next = null;
                        }
                        else
                        {
                            previous.Next = current.Next;
                            next.Prev = current.Prev;

                        }
                        count--;
                    }

                    current = current.Next;
                }                
            }
            return check;
        }

        /// <summary>
        /// Returns true if the list contains the specified item,
        /// false otherwise.
        /// </summary>
        public bool Check(T item)
        {
            DoublyLinkedListNode<T> current = head;
            while (current != null)
            {
                if (current.Value.Equals(item))
                {
                    return true;
                }

                current = current.Next;
            }
            return false;
        }

        /// <summary>
        /// Returns item if the list contains the specified item
        /// </summary>
        /// <returns>Returns item if found</returns>
        public T Search(T item)
        {
            if (!Check(item))
            {
                throw new InvalidOperationException("Item not found in list.");
            }

            DoublyLinkedListNode<T> current = head;
            while (current != null)
            {
                if (current.Value.Equals(item))
                {
                    return current.Value;
                }

                current = current.Next;
            }

            return item;
        }

        /// <summary>
        /// Returns true if the list is empty, otherwise false.
        /// </summary>
        public bool isEmpty()
        {
            return count == 0;
        }

        /// <summary>
        /// Checks for the length of list
        /// </summary>
        /// <returns>Length of list</returns>
        public int Size()
        {
            return count;
        }

        /// <summary>
        /// Finds the index of the specified item
        /// </summary>
        /// <param name="item"></param>
        /// <returns>the index of item</returns>
        public int Index(T item)
        {
            if (!Check(item))
            {
                throw new InvalidOperationException("Item not in list");
            }

            DoublyLinkedListNode<T> current = head;
            var currentIndex = 0;

            while (currentIndex != count)
            {
                if (current.Value.Equals(item))
                {
                    return currentIndex;
                }

                current = current.Next;
                currentIndex++;
            }
            return -1;
        }

        /// <summary>
        /// Length of list, zero if list is empty
        /// </summary>
        public int count
        {
            get;
            private set;
        }

        /// <summary>
        /// The first node in the list or null if empty
        /// </summary>
        private DoublyLinkedListNode<T> head
        {
            get;
            set;
        }

        /// <summary>
        /// The last node in the list or null if empty
        /// </summary>
        private DoublyLinkedListNode<T> tail
        {
            get;
            set;
        }
    }
}

File: Python/DataStructures/DoublyLinkedList/doubly_linked_list.py
class Node:
    def __init__(self, data=None, prev=None, next=None):
        self.data = data
        self.prev = prev
        self.next = next


class DoublyLinkedList:
    def __init__(self):
        self.head = None

    def insert_at_beginning(self, data):
        if self.head is None:
            node = Node(data)
            self.head = node
        else:
            new_head = Node(data, next=self.head)
            self.head.prev = new_head
            self.head = new_head

    def length(self):
        itr = self.head
        count = 0
        while itr:
            count += 1
            itr = itr.next
        return count

    def insert_at_end(self, data):
        if self.head is None:
            self.insert_at_beginning(data)
        else:
            itr = self.head
            while itr.next:
                itr = itr.next
            itr.next = Node(data, prev=itr)

    def insert_at(self, data, at):
        if at == 0:
            self.insert_at_beginning(data)
        elif at == self.length() - 1 and self.length() > 2:
            self.insert_at_end(data)
        else:
            itr = self.head
            count = 0
            while itr:
                if count != at - 1:
                    count += 1
                    itr = itr.next
                else:
                    break
            itr.next = Node(data, itr, itr.next)

    def remove_val(self, val):
        while self.head and self.head.data is val:
            self.head = self.head.next
        if self.head is None:
            return

        current = self.head
        while current:
            if current.data == val:
                if current == self.head:
                    self.head = current.next
                    self.head.prev = None
                if current.next is None:
                    current.prev.next = None
                    return
                else:
                    current.next.prev = current.prev
                    current.prev.next = current.next
            current = current.next

    def __str__(self):
        current = self.head
        as_string = ''
        while current:
            as_string += str(current.data)
            if current.next:
                as_string += ' --> '
            current = current.next
        return as_string
